<a href="https://colab.research.google.com/github/juanpalominoh/Fundamentos-de-Econometria/blob/main/Examen_Final_Heterocedasticidad.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importar base

In [ ]:
library(readxl)

In [ ]:
base <- read_excel("base_anemia.xlsx")

In [ ]:
names(base)

[1] "hhid"         "hvidx"        "bord"         "hwidx"        "sexo"        
 [6] "pesoalnacer"  "edadmeses"    "hemoglobina"  "anemia"       "educmadre"   
[11] "area"         "mad_primaria" "mad_secund"   "mad_superior"

# Regresión OLS

In [ ]:
options(prompt="R> ", digits=4, scipen=999)

In [ ]:
ols <- lm(hemoglobina ~ sexo + pesoalnacer + edadmeses + mad_secund + mad_superior + area, 
          data = base)
summary(ols)


Call:
lm(formula = hemoglobina ~ sexo + pesoalnacer + edadmeses + mad_secund + 
    mad_superior + area, data = base)

Residuals:
   Min     1Q Median     3Q    Max 
-47.31  -6.22   0.31   6.58  43.84 

Coefficients:
              Estimate Std. Error t value             Pr(>|t|)    
(Intercept)  103.14602    0.64412  160.14 < 0.0000000000000002 ***
sexoMujer      0.78933    0.18817    4.19      0.0000275126340 ***
pesoalnacer    0.23889    0.17355    1.38                0.169    
edadmeses      0.23866    0.00591   40.42 < 0.0000000000000002 ***
mad_secund     0.55202    0.26495    2.08                0.037 *  
mad_superior   3.04918    0.29462   10.35 < 0.0000000000000002 ***
areaUrbano     1.56060    0.22490    6.94      0.0000000000042 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.97 on 11357 degrees of freedom
  (433 observations deleted due to missingness)
Multiple R-squared:  0.149,	Adjusted R-squared:  0.149 
F-statistic:  33

# Test White 

In [ ]:
install.packages("lmtest")
library(lmtest)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘zoo’


Loading required package: zoo


Attaching package: ‘zoo’


The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric




In [ ]:
prueba_white <- bptest(ols,
                      ~ sexo + pesoalnacer + edadmeses + mad_secund + mad_superior + area +
                        (sexo*pesoalnacer) + (sexo*edadmeses) + (sexo*mad_secund) + (sexo*mad_superior) + (sexo*area) + 
                        I(pesoalnacer^2) + (pesoalnacer*edadmeses) + (pesoalnacer*mad_secund) + (pesoalnacer*mad_superior) + (pesoalnacer*area) +
                        I(edadmeses^2) + (edadmeses*mad_secund) + (edadmeses*mad_superior) + (edadmeses*area) +
                        (mad_secund*mad_superior) + (mad_secund*area) + 
                        (mad_superior*area),
                        data = base)
print(prueba_white)


	studentized Breusch-Pagan test

data:  ols
BP = 239, df = 22, p-value <0.0000000000000002



Con un valor del test de White de 239 y un nivel de significancia del 1%, se rechaza la H0 de homocedasticidad. Por lo tanto, existe heterocedasticidad en el modelo.


# Test Breusch-Pagan/Godfrey

In [ ]:
bpg <- bptest(hemoglobina ~ sexo + pesoalnacer + edadmeses + mad_secund + mad_superior + area, 
              studentize = F, data=base)
bpg


	Breusch-Pagan test

data:  hemoglobina ~ sexo + pesoalnacer + edadmeses + mad_secund + mad_superior +     area
BP = 233, df = 6, p-value <0.0000000000000002


Con un valor del test de BP de 233 y un nivel de significancia del 1%, se rechaza la H0 de homocedasticidad. Por lo tanto, existe heterocedasticidad en el modelo.
